Numa placement manager

normanrz opened the pull request:
This PR adds the `NUMAPlacementManager` which is able to dynamically migrate chunks based on load heuristics. It is an opt-in feature which is not integrated anywhere yet.

Brief summary of how it works:
- Chunks now have an optional `AccessCounter` which stores access duration information of the chunk. Chunks that are meant to be persistent should carry the access counter.
- This access counter info is populated with the new `ProxyChunk` objects that track the access time using `rdtsc` and the RAII pattern. So far this is only in use for `TableScan` operations.
- The `ChunkMetricsCollectionTask` periodically digests these access counters into a ring buffer for later reference.
- The `MigrationPreparationTask` runs an algorithm that collects the access info of all chunks in the `StorageManager`, calculates a load imbalance metric between the NUMA nodes and outputs a list of chunk migrations that should reduce the load imbalance if applicable.
- These chunks can then be migrated using the `ChunkMigrationTask`. Only completed tasks are considered for migration.
- All of this functionality is only available when `OPOSSUM_NUMA_SUPPORT` is true.

Other changes:
- Removes unnecessary default `NUMAMemoryResource`. One less static object to worry about
- Limits the number of worker threads per `ProcessingUnit` to 200 (which is an arbitrary number)
- Adds an `Unstealable` task queue to the scheduler
- Adds a `PausableLoopThread` class

This PR is not done yet, but I'd still like some feedback already. Some notes that I'd like feedback on:
- There are still some debug outputs that I'll get rid of.
- I can add tests, but haven't thought of a good test case yet that doesn't run forever.
- There are almost no new comments. Please let me know which sections you think should have comments.
- `NUMAPlacementManager` is a null object when `OPOSSUM_NUMA_SUPPORT` is false. Alternatively, we could also remove it entirely, when NUMA support is unavailable.
- Should we add chunk access tracking for other operators as part of this PR?

mrks commented on src/lib/tasks/chunk_metrics_collection_task.cpp:
```diff
@@ -0,0 +1,32 @@
+
+namespace opossum {
+
+ChunkMetricsCollectionTask::ChunkMetricsCollectionTask() {}
```
`= default`?

mrks commented on src/lib/tasks/chunk_migration_task.cpp:
```diff
@@ -0,0 +1,66 @@
+
+namespace opossum {
+
+ChunkMigrationTask::ChunkMigrationTask(const std::string& table_name, const ChunkID chunk_id, int node_id)
```
I think I would only provide one interface. If the caller has only one ChunkID, it should use {}.

mrks commented on src/lib/tasks/chunk_migration_task.cpp:
```diff
@@ -0,0 +1,66 @@
+    : _table_name(table_name), _node_id(node_id), _chunk_ids({chunk_id}) {}
+
+ChunkMigrationTask::ChunkMigrationTask(const std::string& table_name, const std::vector<ChunkID>& chunk_ids,
+                                       int node_id)
```
`target_node_it`?

mrks commented on src/lib/tasks/chunk_migration_task.cpp:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  for (auto chunk_id : _chunk_ids) {
+    if (chunk_id >= table->chunk_count()) {
```
`DebugAssert(chunk_id < table->chunk_count(), "Chunk with given ID does not exist.");`

mrks commented on src/lib/tasks/chunk_migration_task.cpp:
```diff
@@ -0,0 +1,66 @@
+
+    auto& chunk = table->get_chunk(chunk_id);
+
+    if (!chunk_is_completed(chunk, table->chunk_size())) {
```
Comment: Why can't incomplete chunks be migrated?

mrks replied:
Assert(chunk_is_completed, ...)

mrks commented on src/lib/tasks/chunk_migration_task.cpp:
```diff
@@ -0,0 +1,66 @@
+      return;
+    }
+
+    std::cout << "Starting migration " << _table_name << " " << chunk_id << std::endl;
```
Remove debug prints

mrks commented on src/lib/tasks/chunk_migration_task.hpp:
```diff
@@ -0,0 +1,37 @@
+  /**
+   * @brief Checks if a chunks is completed
+   *
+   * See class comment for further explanation
```
Which comment?

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+namespace opossum {
+
+template <class T>
+void print_vector(const std::vector<T>& vec, std::string prefix = "", std::string sep = " ") {
```
Remove debug helper

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  std::cout << std::endl;
+}
+
+struct HotNodeSet {
```
The structs and methods in this file could all use a short comment

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  ChunkID id;
+  int node;
+  size_t byte_size;
+  double hottness;
```
`hotness`? But even better: `temperature`?

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  std::vector<NodeID> hot_nodes;
+  std::vector<NodeID> cold_nodes;
+
+  for (NodeID i = NodeID(0); static_cast<size_t>(i) < node_chunk.size(); i++) {
```
A short description of the algorithm and its result would be great

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+          .cold_nodes = cold_nodes};
+}
+
+std::vector<double> get_node_hottness(const std::vector<ChunkInfo>& chunk_infos, size_t node_count) {
```
Methods like these are really hard to understand without a high-level comment

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  return scale(node_hottness);
+}
+
+bool chunk_is_completed(const Chunk& chunk, const uint32_t max_chunk_size) {
```
Feels like I've seen this before

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  return -1;
+}
+
+std::vector<ChunkInfo> find_hot_chunks(const StorageManager& storage_manager, const std::chrono::milliseconds& lookback,
```
This method could also use some comments

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+}
+
+std::vector<size_t> count_chunks_by_node(const std::vector<ChunkInfo>& chunk_infos, size_t node_count) {
+  std::vector<size_t> result(node_count);
```
What do we need the `node_count` parameter for if it appears to be equal to `chunk_infos.size()`?

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+
+MigrationPreparationTask::MigrationPreparationTask(const NUMAPlacementManagerOptions& options) : _options(options) {}
+
+void MigrationPreparationTask::_on_execute() {
```
comments would be great

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  print_vector(hot_nodes.node_hottness, "Hotnesses: ");
+  print_vector(count_chunks_by_node(hot_chunks, topology->nodes().size()), "Chunk counts: ");
+
+  if (hot_nodes.imbalance > _options.imbalance_threshold && hot_nodes.cold_nodes.size() > 0) {
```
Why does `hot_nodes` have a member `cold_nodes`?

mrks commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,229 @@
+  if (hot_nodes.imbalance > _options.imbalance_threshold && hot_nodes.cold_nodes.size() > 0) {
+    std::vector<ChunkInfo> migration_candidates;
+    for (const auto& hot_chunk : hot_chunks) {
+      if (hot_chunk.node < 0 || contains(hot_nodes.hot_nodes, static_cast<NodeID>(hot_chunk.node))) {
```
Why could `hot_chunk.node < 0`?

normanrz replied:
I use `-1` to identify nodes that are not pinned to a specific node. I could turn that into a named constant.

mrks replied:
That would be easier to understand, yes

mrks commented on src/lib/utils/pausable_loop_thread.hpp:
```diff
@@ -0,0 +1,26 @@
+
+namespace opossum {
+
+struct PausableLoopThread {
```
Could we get short comments here as well?

normanrz commented on src/lib/scheduler/processing_unit.cpp:
```diff
@@ -6,6 +6,8 @@
 #include "uid_allocator.hpp"
 #include "worker.hpp"
 
+#define MAX_WORKERS_PER_CORE 200
```
I picked the number arbitrarily (it worked quite well in my experiments). Not sure how to systematically pick the best number. I easily surpassed it. When you create a lot of table scan operations each task spawns subtasks and waits for them. Now, the original thread is blocked due to waiting. The scheduler now spawns a new thread per subtask (because the original is blocked). If you schedule enough table scans, you'll easily surpass 200 and eventually the whole process becomes overloaded.

Not really sure how to solve this with the current task model other than to limit the number of workers. An async model with a single thread per core and task yielding might solve this.

mrks replied:
This is a completely unrelated tuning issue. Interesting to see that you had more than 200 workers running. Maybe just add a comment that this number is arbitrarily chosen but has been reached in benchmarks.

normanrz commented on src/lib/storage/base_attribute_vector.hpp:
```diff
@@ -26,5 +28,7 @@ class BaseAttributeVector : private Noncopyable {
   // returns the width of the values in bytes
   virtual AttributeVectorWidth width() const = 0;
+
+  virtual std::shared_ptr<BaseAttributeVector> migrate(const PolymorphicAllocator<size_t>& alloc) const = 0;
```
Yes, it is technically a copy. My migration model is: Keep the chunk object, but copy all subordinate objects (columns, attribute vectors) to a new NUMA node.

mrks replied:
Can we come up with a better name then? "migrate" sounds as if the data is gone in the original place.

normanrz replied:
How about `copy_with_allocator` or just `copy`?

mrks replied:
I like `copy_using_allocator` a tad better, because it makes it clear that the allocator is not being copied. Also, a short comment would be great that this can be used to migrate things to a different NUMA node if that's what the other allocator does.

normanrz commented on src/lib/storage/dictionary_column.cpp:
```diff
@@ -21,6 +21,11 @@ DictionaryColumn<T>::DictionaryColumn(const pmr_vector<T>&& dictionary,
     : _dictionary(std::make_shared<pmr_vector<T>>(std::move(dictionary))), _attribute_vector(attribute_vector) {}
 
 template <typename T>
+DictionaryColumn<T>::DictionaryColumn(const std::shared_ptr<pmr_vector<T>>& dictionary,
```
We need this for `DictionaryColumn::migrate` where a new shared_ptr is created. If we only pass the dereferenced shared_ptr it would be deallocated when it goes out of scope (ie. immediately).

normanrz commented on src/lib/storage/numa_placement_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+
+const std::shared_ptr<NUMAPlacementManager>& NUMAPlacementManager::get() { return _instance; }
+
+void NUMAPlacementManager::set(const std::shared_ptr<NUMAPlacementManager>& instance) { _instance = instance; }
```
Wouldn't that make testing more difficult? I copied this approach from StorageManager

mrks replied:
Hmm? StorageManager doesn't have a set method. It does have a reset method that makes testing easy.

normanrz commented on src/lib/storage/numa_placement_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+NUMAPlacementManager::NUMAPlacementManager(const std::shared_ptr<Topology> topology,
+                                           const NUMAPlacementManagerOptions options)
+    : _topology(topology), _options(options) {
+  for (size_t i = 0; i < _topology->nodes().size(); i++) {
```
Actually, I only need `i` for the memory_resource name below.

mrks replied:
ok

normanrz commented on src/lib/storage/numa_placement_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+    : _topology(topology), _options(options) {
+  for (size_t i = 0; i < _topology->nodes().size(); i++) {
+    char msource_name[8];
+    std::snprintf(msource_name, sizeof(msource_name), "numa_%03lu", i);
```
This is not printing, but creating a string. `snprintf` is like `sprintf` but safer. The memory resources from pgasus require a name for some reason. We could just use a single name, but I thought a node-specific makes more sense.

mrks replied:
My fault. A quick comment that pgasus wants this would be great.

normanrz commented on src/lib/storage/numa_placement_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+  collector_thread = std::make_unique<PausableLoopThread>(_options.counter_history_interval, [](size_t) {
+    const auto task = std::make_shared<ChunkMetricsCollectionTask>();
+    task->schedule();
+    task->join();
```
I could also just execute the task. But I thought it may be nicer to schedule it with the scheduler.

mrks replied:
If I am not mistaken, this leads to two threads being used. One for the PausableLoopThread (native) and one for the ChunkMetricsCollectionTask (scheduled). If this makes sense in this case, this needs a comment.

normanrz commented on src/lib/operators/table_scan.cpp:
```diff
@@ -100,10 +101,11 @@ std::shared_ptr<const Table> TableScan::_on_execute() {
       const auto matches_out = std::make_shared<PosList>(_impl->scan_chunk(chunk_id));
 
-      Chunk chunk_out;
+      Chunk chunk_out(chunk_guard->get_allocator(), chunk_guard->access_counter());
```
Yes, the idea is that accesses of derived chunks are still counted towards the original chunk.

mrks replied:
Could you add that as a comment?

normanrz commented:
Thanks for the thorough review @mrks! I will rework the code in the coming days (and add more descriptive code comments).

normanrz commented on src/lib/storage/numa_placement_manager.cpp:
```diff
@@ -0,0 +1,95 @@
+
+std::shared_ptr<NUMAPlacementManager> NUMAPlacementManager::_instance;
+
+const std::shared_ptr<NUMAPlacementManager>& NUMAPlacementManager::get() { return _instance; }
```
I'm having trouble implementing this. I pushed a non-building version on https://github.com/hyrise/hyrise/compare/numa-placement-manager...numa-placement-manager-singleton. Perhaps you could help me with that.

<details>
<summary>Compiler Error</summary>

```
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp: In static member function 'static void opossum::NUMAPlacementManager::reset(std::shared_ptr<opossum::Topology>, opossum::NUMAPlacementManager::Options)':
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:28:49: error: use of deleted function 'opossum::NUMAPlacementManager& opossum::NUMAPlacementManager::operator=(const opossum::NUMAPlacementManager&)'
   get() = NUMAPlacementManager(topology, options);
                                                 ^
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note: 'opossum::NUMAPlacementManager& opossum::NUMAPlacementManager::operator=(const opossum::NUMAPlacementManager&)' is implicitly deleted because the default definition would be ill-formed:
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: no matching function for call to 'std::shared_ptr<opossum::Topology>::operator=(const std::shared_ptr<opossum::Topology>&) const'
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:296:19: note: candidate: std::shared_ptr<_Tp>& std::shared_ptr<_Tp>::operator=(const std::shared_ptr<_Tp>&) [with _Tp = opossum::Topology] <near match>
       shared_ptr& operator=(const shared_ptr&) noexcept = default;
                   ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:296:19: note:   passing 'const std::shared_ptr<opossum::Topology>*' as 'this' argument discards qualifiers
/usr/include/c++/7/bits/shared_ptr.h:300:2: note: candidate: std::shared_ptr<_Tp>::_Assignable<const std::shared_ptr<_Yp>&> std::shared_ptr<_Tp>::operator=(const std::shared_ptr<_Yp>&) [with _Yp = opossum::Topology; _Tp = opossum::Topology; std::shared_ptr<_Tp>::_Assignable<const std::shared_ptr<_Yp>&> = std::shared_ptr<opossum::Topology>&] <near match>
  operator=(const shared_ptr<_Yp>& __r) noexcept
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:300:2: note:   passing 'const std::shared_ptr<opossum::Topology>*' as 'this' argument discards qualifiers
/usr/include/c++/7/bits/shared_ptr.h:309:2: note: candidate: template<class _Yp> std::shared_ptr<_Tp>::_Assignable<std::auto_ptr<_Up> > std::shared_ptr<_Tp>::operator=(std::auto_ptr<_Up>&&) [with _Yp = _Yp; _Tp = opossum::Topology]
  operator=(auto_ptr<_Yp>&& __r)
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:309:2: note:   template argument deduction/substitution failed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note:   types 'std::auto_ptr<_Up>' and 'const std::shared_ptr<opossum::Topology>' have incompatible cv-qualifiers
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:317:7: note: candidate: std::shared_ptr<_Tp>& std::shared_ptr<_Tp>::operator=(std::shared_ptr<_Tp>&&) [with _Tp = opossum::Topology] <near match>
       operator=(shared_ptr&& __r) noexcept
       ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:317:7: note:   conversion of argument 1 would be ill-formed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: cannot bind rvalue reference of type 'std::shared_ptr<opossum::Topology>&&' to lvalue of type 'const std::shared_ptr<opossum::Topology>'
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:325:2: note: candidate: template<class _Yp> std::shared_ptr<_Tp>::_Assignable<std::shared_ptr<_Yp> > std::shared_ptr<_Tp>::operator=(std::shared_ptr<_Yp>&&) [with _Yp = _Yp; _Tp = opossum::Topology]
  operator=(shared_ptr<_Yp>&& __r) noexcept
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:325:2: note:   template argument deduction/substitution failed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note:   types 'std::shared_ptr<_Tp>' and 'const std::shared_ptr<opossum::Topology>' have incompatible cv-qualifiers
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:333:2: note: candidate: template<class _Yp, class _Del> std::shared_ptr<_Tp>::_Assignable<std::unique_ptr<_Up, _Ep> > std::shared_ptr<_Tp>::operator=(std::unique_ptr<_Up, _Ep>&&) [with _Yp = _Yp; _Del = _Del; _Tp = opossum::Topology]
  operator=(unique_ptr<_Yp, _Del>&& __r)
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:333:2: note:   template argument deduction/substitution failed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note:   types 'std::unique_ptr<_Tp, _Dp>' and 'const std::shared_ptr<opossum::Topology>' have incompatible cv-qualifiers
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: passing 'const opossum::NUMAPlacementManager::Options' as 'this' argument discards qualifiers [-fpermissive]
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:21:10: note:   in call to 'constexpr opossum::NUMAPlacementManager::Options& opossum::NUMAPlacementManager::Options::operator=(const opossum::NUMAPlacementManager::Options&)'
   struct Options {
          ^~~~~~~
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = opossum::PausableLoopThread; _Dp = std::default_delete<opossum::PausableLoopThread>]'
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:389:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^~~~~~~~
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = opossum::PausableLoopThread; _Dp = std::default_delete<opossum::PausableLoopThread>]'
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:389:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^~~~~~~~
src/lib/CMakeFiles/hyrise.dir/build.make:2486: recipe for target 'src/lib/CMakeFiles/hyrise.dir/storage/numa_placement_manager.cpp.o' failed
make[2]: *** [src/lib/CMakeFiles/hyrise.dir/storage/numa_placement_manager.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:930: recipe for target 'src/lib/CMakeFiles/hyrise.dir/all' failed
make[1]: *** [src/lib/CMakeFiles/hyrise.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

</details>

mrks replied:
Could you share the diff that leads to this error?

normanrz replied:
Isn't that https://github.com/hyrise/hyrise/compare/numa-placement-manager...numa-placement-manager-singleton ?

mrks replied:
D'oh. I only expanded the error, but missed that link. Sorry.

mrks replied:
If you really need that `reset` method (do you?), you will have to allow the move assignment constructor which was implicitly deleted when you set the other constructors.

normanrz commented:
I implemented the PR feedback, added a bunch of comments to the code and added test cases (that should run fairly fast) for the NUMA chunk migration algorithm.

normanrz commented:
@mrks How do you want to proceed with this? I'm happy to solve the merge conflicts, if you want to move this forward.

mrks commented:
If you could look at those two small things, that would be great. After that, we should be done. Thanks again for your time!

lanice commented:
Somehow, when I'm running the tests on my machine, every time when Topology::create_fake_numa_topology is called, when it returns, it throws this error:

`terminate called after throwing an instance of 'std::out_of_range'`
`what():  vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)`

This does not happen on the pella for example. Anybody an idea what happens here?

normanrz commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,215 @@
+                                           /* .id = */ i,
+                                           /* .node = */ MigrationPreparationTask::get_node_id(chunk.get_allocator()),
+                                           /* .byte_size = */ chunk.byte_size(),
+                                           /* .temperature = */ temperature});
```
Same here :-)

normanrz commented:
@lanice can you try again now? i pushed a fix in https://github.com/hyrise/hyrise/pull/353/commits/505503c1a08c3f54f5653b384c7f879c5f3408d5

normanrz commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,215 @@
+  std::vector<NodeID> cold_nodes;
+
+  for (NodeID i = NodeID(0); static_cast<size_t>(i) < node_chunk.size(); i++) {
+    double a = node_chunk[i];
```
I renamed to `node_temperature` which should be more suitable

normanrz commented on src/lib/tasks/migration_preparation_task.cpp:
```diff
@@ -0,0 +1,215 @@
+    // Identify migration candidates (chunks)
+    std::vector<ChunkInfo> migration_candidates;
+    for (const auto& hot_chunk : hot_chunks) {
+      if (hot_chunk.node < 0 || contains(node_info.hot_nodes, static_cast<NodeID>(hot_chunk.node))) {
```
I added a constant in `NUMAMemoryResource::UNDEFINED_NODE_ID`

mrks commented:
Are you sure that's the problem? Something is trying to access some vector at the *second* position. @lanice - It would be really helpful to get a stack trace for things like this.

normanrz commented:
you're right. https://github.com/hyrise/hyrise/pull/353/commits/505503c1a08c3f54f5653b384c7f879c5f3408d5 still seems like a reasonable change to me

mrks commented:
Isn't that already done here?

https://github.com/hyrise/hyrise/blob/a78220bc7992fbc85c8657c75b5c738c0f620b9c/src/lib/scheduler/topology.cpp#L21

normanrz commented:
Fair enough. I reverted the change

lanice commented:
That's what I get out of gdb:
```
[ RUN      ] OperatorsImportCsvTest.Parallel
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)
terminate called recursively
[New Thread 0x7ffff24c5700 (LWP 4201)]

Thread 2 "hyriseTest" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff2cc6700 (LWP 4199)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff6590f5d in __GI_abort () at abort.c:90
#2  0x00007ffff6f37095 in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff6f34c86 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff6f34cd1 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff6f6112e in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff746d7fc in start_thread (arg=0x7ffff2cc6700) at pthread_create.c:465
#7  0x00007ffff666cb0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

mrks commented:
That is just the teardown of the default exception handler. You will have to catch exceptions so that you get the stack trace of the place where the error happens.

I believe you need `--gtest_catch_exceptions=0`. See the googletest documentation for this: 
https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md

lanice commented:
I'm having difficulties pinning it down. It must occur somewhere in the parallel executed code, which makes it hard to catch exceptions. I think it happens in `ProcessingUnit::wake_or_create_worker`, during these lines:
```
auto fn = std::bind(&Worker::operator(), worker.get());
_threads.emplace_back(fn);
```

mrks commented on src/lib/scheduler/worker.cpp:
```diff
@@ -72,7 +72,9 @@ void Worker::operator()() {
       auto work_stealing_successful = false;
-      for (auto& queue : scheduler->queues()) {
+      const auto queues = {scheduler->queues().at(static_cast<size_t>(SchedulePriority::High)),
+                           scheduler->queues().at(static_cast<size_t>(SchedulePriority::Normal))};
```
This is the part that is causing the problems for Leander. I have no idea what is supposed to be happening here, because it completely contradicts the way the queues are set up:

```
void NodeQueueScheduler::begin() {
  // [...]
  _queues.reserve(_topology->nodes().size());

  for (NodeID q{0}; q < _topology->nodes().size(); q++) {
    auto queue = std::make_shared<TaskQueue>(q);
    _queues.emplace_back(queue);
}
```

If we enable HYRISE_NUMA_SUPPORT but set `max_num_threads` in topology.cpp:16 to 1, the tests should still work.

I would try to fix this myself, but I am not sure of the expected behavior.

normanrz replied:
Good catch, that code was intended to prevent work stealing from `Unstealable` queue, but ended up in the wrong spot. I attempted a fix in f94b398

mrks commented:
Sehr cool! Willst du den Knopf jetzt drücken? :)

normanrz commented:
Yay :tada:
